WT-3181 Add timestamp support.

michaelcahill opened the pull request:
MongoDB needs to control the logical order of operations and read as of
timestamps that are meaningful to the (distributed) system.

michaelcahill commented:
Still lots of work to do here, pushing a branch so this change doesn't live only on my development box.

agorrod commented on src/include/btmem.h:
```diff
@@ -208,7 +208,7 @@ struct __wt_ovfl_txnc {
 #define	WT_LAS_FORMAT							\
     "key_format=" WT_UNCHECKED_STRING(IuQQu)				\
-    ",value_format=" WT_UNCHECKED_STRING(QIu)
+    ",value_format=" WT_UNCHECKED_STRING(QuIu)
```
The comment describing the layout needs to be updated as well.

agorrod commented on src/include/btmem.h:
```diff
@@ -719,6 +719,9 @@ struct __wt_page {
  */
 struct __wt_page_deleted {
 	uint64_t txnid;			/* Transaction ID */
+#if TIMESTAMP_SIZE > 0
```
It's probably no prettier, but does it make sense to abstract these three lines into another macro `WT_DECL_TIMESTAMP`, possibly with a variable name for a parameter?

agorrod commented on src/include/txn.h:
```diff
@@ -88,11 +98,25 @@ struct __wt_txn_global {
+	/* Protects the leading edge of the active transaction window. */
+	WT_RWLOCK current_rwlock;
+
+#if 0
```
I guess you've been doing performance testing here and need to make a choice?

michaelcahill replied:
No, I didn't get it working the other way.  I either need to make it work with two locks (my preference as long as they don't end up entangled), or go back to having just one.

Either way, I'm expecting some followup work to improve concurrency in the transaction subsystem.  In particular, I'm considering moving away from the model where we have a static array of `WT_TXN_STATE` structures that we have to scan regardless of whether the associated session is active, and instead have transactions hook themselves into one/more sorted lists when they start a transaction or pin a transaction ID or timestamp.

My plan is for that work to happen in a followup ticket, after these changes are merged.

agorrod commented on src/include/txn.i:
```diff
@@ -72,8 +103,15 @@ __wt_txn_modify(WT_SESSION_IMPL *session, WT_UPDATE *upd)
+		    WT_TXN_OP_INMEM : WT_TXN_OP_BASIC_TS;
+	} else
+#endif
+		op->type = F_ISSET(session, WT_SESSION_LOGGING_INMEM) ?
```
Indenting looks wrong?

agorrod commented on src/include/wiredtiger.in:
```diff
@@ -2109,6 +2131,48 @@ struct __wt_connection {
+	 * @errors
+	 */
+	int __F(query_timestamp)(
+	    WT_CONNECTION *connection, char *timestamp, const char *config);
```
Did you think about abstracting the `char *timestamp` in a structure? It's referred to in our code as `char *` and `uint8_t` interchangeably, and the length is expected to always be the same, but is wrapped in the `TIMESTAMP_SIZE` macro. It'd be nice if we could refer to them as `transaction_timestamp` (or something) everywhere instead.

michaelcahill replied:
I've created a type for internal, raw timestamps (which were consistently `uint8_t` arrays before.  The `char *` buffers are hexadecimal timestamp strings.  I've changed parameter names to clarify.

agorrod commented on src/txn/txn.c:
```diff
@@ -375,10 +375,260 @@ __wt_txn_update_oldest(WT_SESSION_IMPL *session, uint32_t flags)
+	txn_global = &S2C(session)->txn_global;
+
+	WT_RET(__wt_config_gets(session, cfg, "get", &cval));
+	if (WT_STRING_MATCH("all_committed", cval.str, cval.len)) {
```
I think it'd be nice if these functions were pulled into a different `txn_timestamp.c` source file, I also think it'd be nice in this particular function to split the config parsing from the logic.

agorrod commented:
autobuild: retest this please

michaelcahill commented:
@agorrod, I believe I've incorporated all of your feedback and fixed some remaining issues.  Can you please take another look?

@keithbostic, can you also review please?

Note that I expect a follow up ticket to improve concurrency through `txn_global`.  The current version has a great big rwlock and still scans for snapshots and to update the oldest ID.  In particular, I'm thinking that transactions will hook themselves into lists (including `commit_timestamp` (usually empty), `pinned_txnid`, `running` and `read_timestamp`), each of which is protected by a separate rwlock.

That can come later, but should improve concurrent performance over this version.

michaelcahill commented on src/evict/evict_lru.c:
```diff
@@ -1852,7 +1852,7 @@ __evict_walk_file(WT_SESSION_IMPL *session,
 		if (modified && txn_global->current != txn_global->oldest_id &&
 		    (mod->last_eviction_id == __wt_txn_oldest_id(session) ||
-		    !__wt_txn_visible_all(session, mod->update_txn)))
+		    !__wt_txn_visible_all(session, mod->update_txn, NULL)))
```
Hmm, the `NULL` here is usually a sign that I've missed a case.  I suspect there's something to fix here.

michaelcahill replied:
On closer inspection, this is a heuristic, so it's fine that it doesn't check a timestamp.

michaelcahill commented on src/include/serial.i:
```diff
@@ -312,11 +312,11 @@ __wt_update_serial(WT_SESSION_IMPL *session, WT_PAGE *page,
 			WT_RET(__wt_txn_update_oldest(session, 0));
 
-			if (!__wt_txn_visible_all(session, txn))
+			if (!__wt_txn_visible_all(session, txn, NULL))
```
This is just an optimization (don't retry obsolete checks if nothing has changed).  I don't think it needs to be fixed.

michaelcahill commented on src/include/misc.h:
```diff
@@ -249,6 +249,18 @@
+#ifdef HAVE_TIMESTAMPS
+#define	WT_TIMESTAMP(x) (x)
+typedef uint8_t wt_timestamp_t[TIMESTAMP_SIZE];
+#define	WT_DECL_TIMESTAMP(x) wt_timestamp_t x;
```
Regarding the semicolon: I couldn't figure out a way to move it outside the macro without breaking the no-timestamp case.  Happy to take suggestions.

michaelcahill commented on src/reconcile/rec_write.c:
```diff
@@ -1372,7 +1403,7 @@ __rec_txn_read(WT_SESSION_IMPL *session, WT_RECONCILE *r,
 		if (vpack != NULL &&
 		    vpack->raw == WT_CELL_VALUE_OVFL_RM &&
-		    !__wt_txn_visible_all(session, min_txn))
+		    !__wt_txn_visible_all(session, min_txn, NULL))
```
On review, I think we need to track the update with the oldest transaction ID, and check if it is `visible_all`.  As long as some update is stable, we don't need to copy the original.

agorrod replied:
Did you mean `timestamp` instead of `transaction ID` above? If so I think I agree.

michaelcahill commented on src/lsm/lsm_cursor.c:
```diff
@@ -539,7 +539,7 @@ retry:	if (F_ISSET(clsm, WT_CLSM_MERGE)) {
 				    chunk->switch_txn;
 				if (__wt_txn_visible_all(
-				    session, chunk->switch_txn))
+				    session, chunk->switch_txn, NULL))
```
@agorrod, let's talk about whether LSM needs to track a timestamp when switching chunks.  I thought not when I was changing the code but it's worth discussing (and documenting) if that's correct.

michaelcahill commented on src/txn/txn_nsnap.c:
```diff
@@ -43,13 +43,13 @@ __nsnap_drop_one(WT_SESSION_IMPL *session, WT_CONFIG_ITEM *name)
 	if (found == TAILQ_FIRST(&txn_global->nsnaph)) {
 		WT_ASSERT(session, !__wt_txn_visible_all(
-		    session, txn_global->nsnap_oldest_id));
+		    session, txn_global->nsnap_oldest_id, NULL));
```
Note that named snapshots and timestamps *can* coexist, but I can't think of a reason why an application would want both.

agorrod commented:
I've switched a couple of the stress testers over to this branch.

keithbostic commented on src/txn/txn_timestamp.c:
```diff
@@ -0,0 +1,299 @@
+	    session, active_timestamp, query_cfg)) != 0)
+		return (ret == WT_NOTFOUND ? 0 : ret);
+
+	if (__wt_timestamp_cmp(oldest_timestamp, active_timestamp) < 0) {
```
I don't think `oldest_timestamp` is ever initialized.

michaelcahill replied:
Thanks, that was introduced by some code shuffling.  It's been fixed since.

keithbostic commented on test/checkpoint/workers.c:
```diff
@@ -202,8 +204,22 @@ real_worker(void)
+				}
+				testutil_check(__wt_snprintf(
+				    config_buf, sizeof(config_buf),
+				    "commit_timestamp=%x", ts));
```
These `%x` formats should be `PRIx64`.

keithbostic commented on src/txn/txn_timestamp.c:
```diff
@@ -0,0 +1,299 @@
+	uint32_t i, session_cnt;
+
+	conn = S2C(session);
+	txn_global = &S2C(session)->txn_global;
```
No need for the second `S2C`, just reads oddly.

keithbostic commented on test/format/ops.c:
```diff
@@ -405,6 +405,42 @@ snap_check(WT_CURSOR *cursor,
+		testutil_check(__wt_snprintf(
+		    config_buf, sizeof(config_buf),
+		    "commit_timestamp=%x", ts));
+		commit_conf = config_buf;
```
These `%x` formats should be `PRIx64`.

michaelcahill replied:
Done.

keithbostic commented on test/format/wts.c:
```diff
@@ -540,6 +541,17 @@ wts_verify(const char *tag)
+		testutil_check(__wt_snprintf(
+		    config_buf, sizeof(config_buf),
+		    "oldest_timestamp=%x", g.timestamp));
+		testutil_check(conn->set_timestamp(conn, config_buf));
```
Another `%x` that should be `PRIx64`.

keithbostic commented:
@michaelcahill, I've been struggling to get comfortable with the string-ified API, and I'm mostly there. I think I still have one concrete suggestion: why not store the hexadecimal value in the `WT_UPDATE` structure as an array of `uint64_t`?

Typedef a `uint64_t` array of size [MAX /  16 + 1]`, then copying, assigning and comparing timestamps becomes a register comparison, not a memcmp.

Or are the instructions the same on modern compilers?

michaelcahill commented:
@keithbostic, `memcmp`and `memcpy` of aligned 8-byte arrays should be a single instruction on modern compilers.  Also we've had discussions about 12-byte timestamps, for example, so an array of `uint64_t` won't work in that case.

keithbostic commented:
> memcmp and memcpy of aligned 8-byte arrays should be a single instruction on modern compilers.

I just checked, and as far as I can tell, `clang 4.0` with `-S -O3` calls `memcmp` to compare aligned, 8-byte arrays.

> Also we've had discussions about 12-byte timestamps, for example, so an array of uint64_t won't work in that case.

A `uint64_t` can hold up to 16 hex bytes, right? So 12-bytes would be fine.

For a 17-byte hex timestamp you'd need a second `uint64_t`, but it still works as long as you compare the `uint64_t` pair that holds the MSB first.

Admittedly, you've wasted some space in that case, but you could also hide something more space efficient behind the timestamp compare/set functions.

michaelcahill commented:
Just to clarify, `TIMESTAMP_SIZE` is the number of bytes of raw timestamp, so `TIMESTAMP_SIZE=8` means that the hex-encoding will be up to 16 bytes plus a NUL byte.

Getting back to the question of execution efficiency, I looked at this piece of code in `txn.c`:

```c
 445                 WT_RET(__wt_txn_parse_timestamp(
 446                     session, "read", txn->read_timestamp, &cval));
 447                 __wt_readlock(session, &txn_global->rwlock);
 448                 __wt_timestamp_set(
 449                     oldest_timestamp, txn_global->oldest_timestamp);
 450                 __wt_readunlock(session, &txn_global->rwlock);
```

With "MongoDB clang version 3.9.1" (i.e., whatever version of the MongoDB toolchain I last downloaded), this turns into:

```
.Ltmp311:
        #DEBUG_VALUE: __wt_txn_config:cval <- [%RCX+0]
        movq    %r14, %rdi
        movq    %r15, %rdx
        callq   __wt_txn_parse_timestamp@PLT
.Ltmp312:
        movl    %eax, %ebx
.Ltmp313:
        #DEBUG_VALUE: __ret <- %EBX
        testl   %ebx, %ebx
        jne     .LBB4_24
.Ltmp314:
# BB#20:
        #DEBUG_VALUE: __ret <- %EBX
        #DEBUG_VALUE: __wt_timestamp_set:src <- %R15
        #DEBUG_VALUE: __wt_timestamp_cmp:ts1 <- %R15
        #DEBUG_VALUE: __wt_txn_config:session <- %R14
        movq    %r13, -80(%rbp)         # 8-byte Spill
        .loc    37 447 39 is_stmt 1     # ../src/txn/txn.c:447:39
        leaq    26728(%r12), %rbx
.Ltmp315:
        .loc    37 447 3 is_stmt 0      # ../src/txn/txn.c:447:3
        movq    %r14, %rdi
        movq    %rbx, %rsi
        callq   __wt_readlock@PLT
.Ltmp316:
        .loc    3 30 8 is_stmt 1        # ../src/include/txn.i:30:8
        movq    26648(%r12), %rax
        movq    %rax, -96(%rbp)
.Ltmp317:
        .loc    37 450 3                # ../src/txn/txn.c:450:3
        movq    %r14, %rdi
        movq    %rbx, %rsi
        callq   __wt_readunlock@PLT
```

Note that between the calls to `_wt_readlock` and `__wt_readunlock`, we're doing direct 8-byte `movq`, not calling to `memcpy`.

Similarly for MongoDB's build of gcc 5.4.0:

```
.LVL237:
        movq    %rax, -104(%rbp)        # D.41029, %sfp
        movl    256(%r15), %eax # session_11(D)->id, D.41025
        movl    %eax, -108(%rbp)        # D.41025, %sfp
.LVL238:
.LBB176:
        .loc 1 445 0
        call    __wt_txn_parse_timestamp@PLT    #
.LVL239:
        testl   %eax, %eax      # D.41020
        movl    %eax, %ebx      #, D.41020
.LVL240:
        jne     .L226   #,
.LBE176:
        .loc 1 447 0
        leaq    26728(%r14), %r12       #, D.41031
        movq    %r15, %rdi      # session,
        movq    %r12, %rsi      # D.41031,
        call    __wt_readlock@PLT       #
.LVL241:
.LBB177:
.LBB178:
        .loc 3 30 0
        movq    26648(%r14), %rax       # MEM[(char * {ref-all})_48 + 26648B], MEM[(char * {ref-all})_48 + 26648B]
.LBE178:
.LBE177:
        .loc 1 450 0
        movq    %r12, %rsi      # D.41031,
        movq    %r15, %rdi      # session,
.LBB180:
.LBB179:
        .loc 3 30 0
        movq    %rax, -96(%rbp) # MEM[(char * {ref-all})_48 + 26648B], MEM[(char * {ref-all})&oldest_timestamp]
.LVL242:
.LBE179:
.LBE180:
        .loc 1 450 0
        call    __wt_readunlock@PLT     #
```

keithbostic commented:
Here's a chunk of code from `__wt_txn_commit`, where it calls `__wt_timestamp_cmp`:
```
.Ltmp390:
        #DEBUG_VALUE: __wt_timestamp_cmp:ts1 <- RDI
        leaq    -80(%rbp), %rsi
.Ltmp391:
        #DEBUG_VALUE: __wt_timestamp_set:dest <- RSI
        #DEBUG_VALUE: __wt_timestamp_cmp:ts2 <- RSI
        movl    $8, %edx
        .loc    4 21 0                  # ./src/include/txn.i:21:0
.Ltmp392:
        callq   memcmp
        testl   %eax, %eax
```

keithbostic commented on src/txn/txn_timestamp.c:
```diff
@@ -0,0 +1,299 @@
+	char padbuf[2 * TIMESTAMP_SIZE];
+	wt_timestamp_t tsbuf;
+	const char *hexts;
+	size_t hexlen;
```
KNF variable order.

keithbostic commented on src/txn/txn_timestamp.c:
```diff
@@ -0,0 +1,299 @@
+		WT_RET_MSG(session, EINVAL,
+		    "Failed to parse %s timestamp '%.*s': too long",
+		    name, (int)cval->len, cval->str);
+
```
Off-by-one if `TIMESTAMP_SIZE` is odd, `cval->len == 2 * TIMESTAMP_SIZE`, we add a byte to the beginning of `padbuf` then copy `cval->len`.

keithbostic commented on src/txn/txn_timestamp.c:
```diff
@@ -0,0 +1,299 @@
+	WT_TXN_GLOBAL *txn_global;
+	wt_timestamp_t active_timestamp;
+	wt_timestamp_t oldest_timestamp;
+	wt_timestamp_t pinned_timestamp;
```
Each getting their own line reads a little oddly, but *shrug*.

keithbostic commented on src/txn/txn_timestamp.c:
```diff
@@ -0,0 +1,299 @@
+ */
+int
+__wt_txn_set_timestamp(
+    WT_SESSION_IMPL *session, const char *cfg[])
```
Unnecessary line break.

michaelcahill commented on src/docs/transactions.dox:
```diff
@@ -148,4 +148,27 @@ WT_SESSION::snapshot with a configuration that includes
+
+Applications can assign transactions explicit commit timestamps, then read
+as-of a timestamp.  Timestamps are communicated to WiredTiger as hexadecimal
+values with a length configurable at build time (default 8 bytes).
```
I see your point, but I don't think applications will care too much about the encoding.  Instead, they will have some native representation of a timestamp with some size (e.g., MongoDB's 8-byte OpTime).  That has to pass through the WiredTiger API somehow, but I don't think we should focus on the size of the encoding -- applications care about how big the actual timestamp can be.

keithbostic replied:
The application doesn't know (or care) how we store a timestamp inside WiredTiger, all the application knows is the maximum size of the string it's going to pass in, so that's what the application can configure.

Imagine an application timestamp that's a maximum of 19 bytes. What calculation should an application writer perform to decide what to use as the "length configurable at build time"?

michaelcahill replied:
If the application's timestamp size is 19 bytes, it should configure a timestamp size of 19.  The encoding would be up to 38 bytes of hex plus a NUL byte.

michaelcahill commented on src/docs/transactions.dox:
```diff
@@ -148,4 +148,27 @@ WT_SESSION::snapshot with a configuration that includes
+
+Commit timestamps cannot be set earlier than the latest read timestamp that has
+been used.  This is enforced by assertions in diagnostic builds, if
+applications violate this rule, inconsistent data can be read.
```
We rely on repeatable reads in various places (e.g., to make sure we don't fail during index reads): if we allow application to violate that, then we can't provide any isolation guarantees.

michaelcahill commented on src/docs/transactions.dox:
```diff
@@ -148,4 +148,27 @@ WT_SESSION::snapshot with a configuration that includes
+future read timestamps will be at least as recent as the oldest timestamp, so
+WiredTiger can discard history before the specified point.  It is critical that
+the oldest timestamp update frequently or the cache can become full of updates,
+reducing performance.
```
Yes, thanks, fixed.

keithbostic commented on src/docs/transactions.dox:
```diff
@@ -149,26 +149,39 @@ the semantics supported by the drop configuration.
+The application's timestamp size is specified as a number of bytes at build
+time, with <code>configure --with-timestamp-size=X</code>.  The default
+timestamp size is 8 bytes (i.e., 64 bits).  Setting a size of zero disables
+transaction timestamp functionality.
```
Is there a performance reason to disable transaction timestamp functionality? (If so, maybe the default should be 0, or we should call that out so applications know to turn it off?)

agorrod replied:
There is a performance cost, though I'm not sure it's measurable. In the least the lock-protected transaction ID management gets a bit more complex with timestamps enabled

michaelcahill replied:
I want the default to be the same as what MongoDB is going to use by default.  We could certainly have a job that runs with it disabled, and I plan to add a configuration test to make sure the build with timestamps disabled continues to work.

michaelcahill commented on src/txn/txn.c:
```diff
@@ -813,11 +973,49 @@ __wt_txn_global_destroy(WT_SESSION_IMPL *session)
+		if (txn_global->oldest_id == txn_global->current &&
+		    txn_global->metadata_pinned == txn_global->current)
+			break;
+		__wt_yield();
```
That loop has always been there, I just moved it from connection code to transaction code.  I don't know what "too long is" -- it might have to wait for a checkpoint to complete, for example.

michaelcahill commented on src/include/btmem.h:
```diff
@@ -207,8 +208,8 @@ struct __wt_ovfl_txnc {
-    "key_format=" WT_UNCHECKED_STRING(IuQQu)				\
-    ",value_format=" WT_UNCHECKED_STRING(QBu)
+    "key_format=" WT_UNCHECKED_STRING(IuQQuu)				\
+    ",value_format=" WT_UNCHECKED_STRING(QuBu)
```
No, the change to the lookaside table is unconditional (it always contains a timestamp field), but the item is always empty when timestamps are disabled.

michaelcahill commented on src/include/btmem.h:
```diff
@@ -906,6 +910,7 @@ struct __wt_ikey {
  */
 WT_PACKED_STRUCT_BEGIN(__wt_update)
 	uint64_t txnid;			/* transaction */
+	WT_DECL_TIMESTAMP(timestamp)
```
The struct is currently padded, and we're expecting timestamps to be 8 bytes in the first real use case (i.e., MongoDB).  We already check that the update structure is the size we expect with a build-time assertion, so I'm inclined to revisit this if we see that assertion fail.  I'm concerned about putting the timestamp *after* the `type` field for the same reason: access to an 8-byte timestamp is more likely to be efficient if it is on an 8 byte boundary.

michaelcahill commented on src/include/wiredtiger.in:
```diff
@@ -2173,6 +2195,52 @@ struct __wt_connection {
+	 */
+	int __F(set_timestamp)(
+	    WT_CONNECTION *connection, const char *config);
+	/*! @} */
```
I deliberately changed from `set_oldest_timestamp` to `set_timestamp("oldest_timestamp=X")` so that we have the flexibility to add future timestamp information.  For example, I expect to reuse this function in the recover to a timestamp project for `set_timestamp("durable_timestamp=X")`

michaelcahill commented on dist/api_data.py:
```diff
@@ -1204,6 +1220,21 @@ def __cmp__(self, other):
+'WT_CONNECTION.set_timestamp' : Method([
+    Config('oldest_timestamp', '', r'''
+        future commits and queries will be no earlier than the specified
+        timestamp. Supplied values must be monotonically increasing.
```
Timestamps have to be mapped into an integral space (which is really the same thing as saying they compare lexicographically).  It's okay (and not surprising) if the same timestamp is specified multiple times, so I think "monotonically increasing" is accurate.

michaelcahill commented on src/btree/bt_delete.c:
```diff
@@ -242,8 +242,10 @@ __wt_delete_page_skip(WT_SESSION_IMPL *session, WT_REF *ref, bool visible_all)
+	    __wt_txn_visible_all(session,
+		ref->page_del->txnid, WT_GET_TIMESTAMP(ref->page_del)):
+	    __wt_txn_visible(session,
+		ref->page_del->txnid, WT_GET_TIMESTAMP(ref->page_del)));
```
Agreed, great idea.  I opened WT-3382 to follow up on this.

michaelcahill commented on src/btree/bt_read.c:
```diff
@@ -18,40 +18,37 @@ int
+	las_key.size = 0;
+	las_timestamp.size = 0;
+	cursor->set_key(cursor, btree_id, &las_addr,
+	    (uint64_t)0, (uint32_t)0, &las_timestamp, &las_key);
```
To be clear, I don't think there was anything wrong before, I'd just like to avoid heap allocation unless we really need it.

michaelcahill commented on dist/api_data.py:
```diff
@@ -1204,6 +1220,21 @@ def __cmp__(self, other):
+        specify which timestamp to query: \c all_committed returns the largest
+        timestamp such that all earlier timestamps have committed.  See @ref
+        transaction_timestamps''',
+        choices=['all_committed']),
```
We don't have a way to list undocumented `choices`.  I've added a comment.

michaelcahill commented on src/txn/txn_timestamp.c:
```diff
@@ -0,0 +1,299 @@
+		}
+		__wt_readunlock(session, &txn_global->rwlock);
+	} else
+		__wt_timestamp_set(ts, zero_timestamp);
```
Agreed, changed to an error rather than being permissive.

michaelcahill commented on src/include/txn.i:
```diff
@@ -128,7 +166,8 @@ __wt_txn_oldest_id(WT_SESSION_IMPL *session)
 	WT_READ_BARRIER();
-	checkpoint_pinned = txn_global->checkpoint_pinned;
+	if (!include_checkpoint_txn)
+		return (oldest_id);
```
Done.

michaelcahill commented on test/suite/test_timestamp01.py:
```diff
@@ -0,0 +1,69 @@
+import random
+from suite_subprocess import suite_subprocess
+import wiredtiger, wttest
+from wtscenario import make_scenarios
```
Yes, we should integrated `pylint` into our checking at some point...

michaelcahill commented on src/txn/txn_ext.c:
```diff
@@ -101,5 +101,5 @@ __wt_ext_transaction_visible(
 
 	return (__wt_txn_visible(
-	    (WT_SESSION_IMPL *)wt_session, transaction_id));
+	    (WT_SESSION_IMPL *)wt_session, transaction_id, NULL));
```
Yes, I've added a section about the extension API after the main part of the timestamp documentation.

michaelcahill commented:
After reviews from @agorrod and @keithbostic, I'm going to go ahead and merge this one today so it gets testing on `develop`.
